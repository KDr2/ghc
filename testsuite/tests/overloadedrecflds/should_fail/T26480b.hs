{-# LANGUAGE AllowAmbiguousTypes #-}
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE RebindableSyntax #-}
{-# LANGUAGE OverloadedRecordDot #-}
{-# LANGUAGE OverloadedRecordUpdate #-}

module T26480b where

import Prelude
import Data.Proxy
import GHC.TypeLits
import GHC.Records


setField
  :: forall (fld :: Symbol) rec ty
  .  HasField fld rec ty => ty -> rec -> rec
setField _ r = r

data N = N { no :: H }

data D = MkD{ field1 :: G }

data G = MkG { xyzzywyzzydyzzy :: H }

data H = MkH { field2 :: Int }

-- Direct usage of 'getField'
test1 :: G -> H
test1 = getField @"xyzzywyzzydyzza"

test1' :: N -> H
test1' = getField @"xyzzywyzzydyzzy"

test1'' :: N -> H
test1'' = getField @"ayzzywyzzydyzzy"

-- Record dot, applied
test2a :: G -> H
test2a g = g.xyzzywyzzydyzzb

test2b :: D -> H
test2b g = g.field1.xyzzywyzzydyzzc

-- Record dot, bare selector
test3a :: G -> H
test3a = (.xyzzywyzzydyzzd)

test3b :: D ->H
test3b = (.field1.xyzzywyzzydyzze)

-- Overloaded record update
test4a :: G -> G
test4a d = d { xyzzywyzzydyzzf = MkG ( MkH 3 ) }

test4b :: D -> D
test4b d = d { field1.xyzzywyzzydyzzg = MkH 3 }
