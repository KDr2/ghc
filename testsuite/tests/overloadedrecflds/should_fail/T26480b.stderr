T26480b.hs:30:9: error: [GHC-39999]
    • No instance for ‘HasField "xyzzywyzzydyzza" G H’
        arising from a use of ‘getField’
      NB: ‘G’ does not have a record field named ‘xyzzywyzzydyzza’.
    • In the expression: getField @"xyzzywyzzydyzza"
      In an equation for ‘test1’: test1 = getField @"xyzzywyzzydyzza"
    Suggested fix: Perhaps use ‘xyzzywyzzydyzzy’ (line 24).

T26480b.hs:33:10: error: [GHC-39999]
    • No instance for ‘HasField "xyzzywyzzydyzzy" N H’
        arising from a use of ‘getField’
      NB: ‘N’ does not have a record field named ‘xyzzywyzzydyzzy’.
      Another type has a field of this name: ‘G’.
    • In the expression: getField @"xyzzywyzzydyzzy"
      In an equation for ‘test1'’: test1' = getField @"xyzzywyzzydyzzy"

T26480b.hs:36:11: error: [GHC-39999]
    • No instance for ‘HasField "ayzzywyzzydyzzy" N H’
        arising from a use of ‘getField’
      NB: ‘N’ does not have a record field named ‘ayzzywyzzydyzzy’.
    • In the expression: getField @"ayzzywyzzydyzzy"
      In an equation for ‘test1''’: test1'' = getField @"ayzzywyzzydyzzy"
    Suggested fix:
      Perhaps use the similarly named field of another type:
        ‘G’: ‘xyzzywyzzydyzzy’ (line 24)

T26480b.hs:40:12: error: [GHC-39999]
    • No instance for ‘HasField "xyzzywyzzydyzzb" G H’
        arising from selecting the field ‘xyzzywyzzydyzzb’
      NB: ‘G’ does not have a record field named ‘xyzzywyzzydyzzb’.
    • In the expression: g.xyzzywyzzydyzzb
      In an equation for ‘test2a’: test2a g = g.xyzzywyzzydyzzb
    Suggested fix: Perhaps use ‘xyzzywyzzydyzzy’ (line 24).

T26480b.hs:43:12: error: [GHC-39999]
    • No instance for ‘HasField "xyzzywyzzydyzzc" G H’
        arising from selecting the field ‘xyzzywyzzydyzzc’
      NB: ‘G’ does not have a record field named ‘xyzzywyzzydyzzc’.
    • In the expression: g.field1.xyzzywyzzydyzzc
      In an equation for ‘test2b’: test2b g = g.field1.xyzzywyzzydyzzc
    Suggested fix: Perhaps use ‘xyzzywyzzydyzzy’ (line 24).

T26480b.hs:47:10: error: [GHC-39999]
    • No instance for ‘HasField "xyzzywyzzydyzzd" G H’
        arising from the record selector ‘xyzzywyzzydyzzd’
      NB: ‘G’ does not have a record field named ‘xyzzywyzzydyzzd’.
    • In the expression: (.xyzzywyzzydyzzd)
      In an equation for ‘test3a’: test3a = (.xyzzywyzzydyzzd)
    Suggested fix: Perhaps use ‘xyzzywyzzydyzzy’ (line 24).

T26480b.hs:50:10: error: [GHC-39999]
    • No instance for ‘HasField "xyzzywyzzydyzze" G H’
      NB: ‘G’ does not have a record field named ‘xyzzywyzzydyzze’.
    • In the expression: (.field1.xyzzywyzzydyzze)
      In an equation for ‘test3b’: test3b = (.field1.xyzzywyzzydyzze)
    Suggested fix: Perhaps use ‘xyzzywyzzydyzzy’ (line 24).

T26480b.hs:54:12: error: [GHC-39999]
    • No instance for ‘HasField "xyzzywyzzydyzzf" G G’
        arising from a record update
      NB: ‘G’ does not have a record field named ‘xyzzywyzzydyzzf’.
    • In the expression: d {xyzzywyzzydyzzf = MkG (MkH 3)}
      In an equation for ‘test4a’:
          test4a d = d {xyzzywyzzydyzzf = MkG (MkH 3)}
    Suggested fix: Perhaps use ‘xyzzywyzzydyzzy’ (line 24).

T26480b.hs:57:12: error: [GHC-39999]
    • No instance for ‘HasField "xyzzywyzzydyzzg" G H’
      NB: ‘G’ does not have a record field named ‘xyzzywyzzydyzzg’.
    • In the expression: d {field1.xyzzywyzzydyzzg = MkH 3}
      In an equation for ‘test4b’:
          test4b d = d {field1.xyzzywyzzydyzzg = MkH 3}
    Suggested fix: Perhaps use ‘xyzzywyzzydyzzy’ (line 24).

